api/envoy/config/common/mutation_rules/v3/mutation_rules.proto

```diff
@@ -0,0 +1,65 @@
+syntax = "proto3";
+
+package envoy.config.common.mutation_rules.v3;
+
+import "envoy/type/matcher/v3/regex.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.config.common.mutation_rules.v3";
+option java_outer_classname = "MutationRulesProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/config/common/mutation_rules/v3;mutation_rulesv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// The HeaderMutationRules structure specifies what headers may be
+// manipulated by a processing filter. By default, all mutations are
+// possible. This rule makes it possible to disallow modifications of
+// certain filters according to the parameters below.
+//
+// This structure is used by the External Processing Filter to make
+// it possible to restrict what an external processing server can do.
+// Every attempt to add, change, append, or remove a header will be
+// tested against the rules here. Disallowed header mutations will be
+// ignored unless "disallow_is_error" is set to true.
+//
+// In addition, a counter will be incremented whenever a mutation is
+// rejected. In the ext_proc filter, that counter is named
+// "rejected_header_mutations".
+// [#next-free-field: 8]
+message HeaderMutationRules {
+  // Prevent modification of headers that could affect routing
+  // decisions. This includes "host", ":authority", ":scheme",
+  // and ":method".
+  bool disallow_routing = 1;
+
+  // Prevent modification of envoy-internal headers. These typically
+  // start with "x-envoy" but this prefix may be overridden elsewhere.
+  bool disallow_envoy = 2;
+
+  // Prevent modification of any system header, defined as a header that
+  // starts with a ":" character.
+  bool disallow_system = 3;
```

Comment:
Should we go for "secure by default here" and make the defaults conservative? That way folks will opt-in to the more dangerous behaviors.

Reply:
In the latest revision, I split the difference between mutations that are allowed by default and disallowed by default to match how ext_proc works today. In that mode, certain changes are disallowed that we agreed previously might cause problems. 

I'm not sure I like the idea of making ALL header mutations disallowed by default, however, as it makes it very difficult to do anything useful with ext_proc at all without changing the defaults. Take a look at what's there now and see if it makes sense.